File: transfer.go
package main

import (
	"flag"
	"fmt"
	"math/rand"
	"os"
	"regexp"
	"runtime"
	"sync/atomic"
	"time"
	"golang.org/x/text/language"
	"golang.org/x/text/message"
)

type LeaderInfo struct {
	id uint64
	idx uint64
	estop uint64
	seed uint64
}

func __xorshift64( state * uint64 ) (uint64) {
	x := *state
	x ^= x << 13
	x ^= x >> 7
	x ^= x << 17
	*state = x
	return x
}

func (this * LeaderInfo) next(len uint64) {
	n := __xorshift64( &this.seed )
	atomic.StoreUint64( &this.id, n % len )
}

func NewLeader(size uint64) (*LeaderInfo) {
	this := &LeaderInfo{0, 0, 0, uint64(os.Getpid())}

	r := rand.Intn(10)

	for i := 0; i < r; i++ {
		this.next( uint64(nthreads) )
	}

	return this
}

type MyThread struct {
	id uint64
	idx uint64
	sem chan struct{}
}

func waitgroup(leader * LeaderInfo, idx uint64, threads [] MyThread, main_sem chan struct {}) {
	start := time.Now()
	Outer:
	for i := 0; i < len(threads); i++ {
		// fmt.Fprintf(os.Stderr, "Waiting for :%d (%d)\n", threads[i].id, atomic.LoadUint64(&threads[i].idx) );
		for atomic.LoadUint64( &threads[i].idx ) != idx {
			// hint::spin_loop();
			end := time.Now()
			delta := end.Sub(start)
			if delta.Seconds() > 5 {
				fmt.Fprintf(os.Stderr, "Programs has been blocked for more than 5 secs")
				atomic.StoreUint64(&leader.estop, 1);
				main_sem <- (struct {}{})
				break Outer
			}
		}
	}
	// debug!( "Waiting done" );
}

func wakegroup(exhaust bool, me uint64, threads [] MyThread) {
	if !exhaust { return; }

	for i := uint64(0); i < uint64(len(threads)); i++ {
		if i != me {
			// debug!( "Leader waking {}", i);
			defer func() {
				if err := recover(); err != nil {
					fmt.Fprintf(os.Stderr, "Panic occurred: %s\n", err)
				}
			}()
			threads[i].sem <- (struct {}{})
		}
	}
}

func lead(exhaust bool, leader * LeaderInfo, this * MyThread, threads [] MyThread, main_sem chan struct {}) {
	nidx := atomic.LoadUint64(&leader.idx) + 1;
	atomic.StoreUint64(&this.idx, nidx);
	atomic.StoreUint64(&leader.idx, nidx);

	if nidx > stop_count || atomic.LoadUint64(&leader.estop) != 0 {
		// debug!( "Leader {} done", this.id);
		main_sem <- (struct {}{})
		return
	}

	// debug!( "====================\nLeader no {} : {}", nidx, this.id);

	waitgroup(leader, nidx, threads, main_sem);

	leader.next( uint64(len(threads)) );

	wakegroup(exhaust, this.id, threads);

	// debug!( "Leader no {} : {} done\n====================", nidx, this.id);
}

func waitleader(exhaust bool, leader * LeaderInfo, this * MyThread, rechecks * uint64) {
	runtime.Gosched()

	if atomic.LoadUint64(&leader.idx) == atomic.LoadUint64(&this.idx) {
		// debug!("Waiting {} recheck", this.id);
		*rechecks += uint64(1)
		return
	}

	// debug!("Waiting {}", this.id);

	// debug_assert!( (leader.idx.load(Ordering::Relaxed) - 1) == this.idx.load(Ordering::Relaxed) );
	atomic.AddUint64(&this.idx, 1)
	if exhaust {
		// debug!("Waiting {} sem", this.id);
		<- this.sem
	} else {
		// debug!("Waiting {} yield", this.id);
		runtime.Gosched()
	}

	// debug!("Waiting {} done", this.id);
}

func transfer_main( result chan uint64, me uint64, leader * LeaderInfo, threads [] MyThread, exhaust bool, start chan struct{}, main_sem chan struct{}) {
	// assert!( me == threads[me].id );

	// debug!("Ready {}: {:p}", me, &threads[me].sem as *const sync::Semaphore);

	// Wait for start
	<- start

	// debug!( "Start {}", me );

	// Prepare results
	r := uint64(0)

	// Main loop
	for true {
		if atomic.LoadUint64(&leader.id) == me {
			lead( exhaust, leader, &threads[me], threads, main_sem )
			runtime.Gosched()
		} else {
			waitleader( exhaust, leader, &threads[me], &r )
		}
		if atomic.LoadUint64(&leader.idx) > stop_count || atomic.LoadUint64(&leader.estop) != 0 { break; }
	}

	// return result
	result <- r
}

func main() {
	// Benchmark specific command line arguments
	exhaustOpt := flag.String("e", "no", "Whether or not threads that have seen the new epoch should park instead of yielding.")

	// General benchmark initialization and deinitialization
	bench_init()

	exhaustVal := *exhaustOpt;

	var exhaust bool
	re_yes := regexp.MustCompile("[Yy]|[Yy][Ee][Ss]")
	re_no  := regexp.MustCompile("[Nn]|[Nn][Oo]")
	if re_yes.Match([]byte(exhaustVal)) {
		exhaust = true
	} else if re_no.Match([]byte(exhaustVal)) {
		exhaust = false
	} else {
		fmt.Fprintf(os.Stderr, "Unrecognized exhaust(-e) option '%s'\n", exhaustVal)
		os.Exit(1)
	}

	if clock_mode {
		fmt.Fprintf(os.Stderr, "Programs does not support fixed duration mode\n")
		os.Exit(1)
	}

	var es string
	if exhaust {
		es = "with"
	} else {
		es = "without"
	}
	fmt.Printf("Running %d threads on %d processors, doing %d iterations, %s exhaustion\n", nthreads, nprocs, stop_count, es );

	main_sem := make(chan struct{})
	leader := NewLeader(uint64(nthreads))
	barr := make(chan struct{})
	result := make(chan uint64)

	thddata := make([]MyThread, nthreads)
	for i := range thddata {
		thddata[i] = MyThread{ uint64(i), 0, make(chan struct {}) }
	}

	rechecks := uint64(0)
	for i := range thddata {
		go transfer_main(result, uint64(i), leader, thddata, exhaust, barr, main_sem)
	}
	fmt.Printf("Starting\n");

	start := time.Now()
	close(barr) // release barrier

	<- main_sem

	end := time.Now()
	delta := end.Sub(start)

	fmt.Printf("\nDone\n")

	// release all the blocked threads
	for i := range thddata {
		close(thddata[i].sem)
	}
	for range thddata {
		rechecks += <- result
	}

	p := message.NewPrinter(language.English)
	var ws string
	if exhaust {
		ws = "yes"
	} else {
		ws = "no"
	}
	p.Printf("Duration (ms)           : %d\n", delta.Milliseconds() )
	p.Printf("Number of processors    : %d\n", nprocs )
	p.Printf("Number of threads       : %d\n", nthreads )
	p.Printf("Total Operations(ops)   : %15d\n", (leader.idx - 1) )
	p.Printf("Threads parking on wait : %s\n", ws)
	p.Printf("Rechecking              : %d\n", rechecks )
	p.Printf("ms per transfer         : %f\n", float64(delta.Milliseconds()) / float64(leader.idx) )
}

File: transfer.rs
#[cfg(debug_assertions)]
use std::io::{self, Write};

use std::process;
use std::option;
use std::hint;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::time::{Instant,Duration};

use tokio::runtime::Builder;
use tokio::sync;
use tokio::task;

use rand::Rng;

use clap::{Arg, App};
use num_format::{Locale, ToFormattedString};

#[path = "../bench.rs"]
mod bench;

#[cfg(debug_assertions)]
macro_rules! debug {
	($x:expr) => {
		println!( $x );
		io::stdout().flush().unwrap();
	};
	($x:expr, $($more:expr),+) => {
		println!( $x, $($more), * );
		io::stdout().flush().unwrap();
	};
}

#[cfg(not(debug_assertions))]
macro_rules! debug {
    ($x:expr   ) => { () };
    ($x:expr, $($more:expr),+) => { () };
}

fn parse_yes_no(opt: option::Option<&str>, default: bool) -> bool {
	match opt {
		Some(val) => {
			match val {
				"yes" => true,
				"Y" => true,
				"y" => true,
				"no"  => false,
				"N"  => false,
				"n"  => false,
				"maybe" | "I don't know" | "Can you repeat the question?" => {
					eprintln!("Lines for 'Malcolm in the Middle' are not acceptable values of parameter 'exhaust'");
					std::process::exit(1);
				},
				_ => {
					eprintln!("parameter 'exhaust' must have value 'yes' or 'no', was {}", val);
					std::process::exit(1);
				},
			}
		},
		_ => {
			default
		},
	}
}

struct LeaderInfo {
	id: AtomicUsize,
	idx: AtomicUsize,
	estop: AtomicBool,
	seed: u128,
}

impl LeaderInfo {
	pub fn new(nthreads: usize) -> LeaderInfo {
		let this = LeaderInfo{
			id: AtomicUsize::new(nthreads),
			idx: AtomicUsize::new(0),
			estop: AtomicBool::new(false),
			seed: process::id() as u128
		};

		let mut rng = rand::thread_rng();

		for _ in 0..rng.gen_range(0..10) {
			this.next( nthreads );
		}

		this
	}

	pub fn next(&self, len: usize) {
		let n = bench::_lehmer64( unsafe {
			let r1 = &self.seed as *const u128;
			let r2 = r1 as *mut u128;
			&mut *r2
		} ) as usize;
		self.id.store( n % len , Ordering::SeqCst );
	}
}

struct MyThread {
	id: usize,
	idx: AtomicUsize,
	sem: sync::Semaphore,
}

fn waitgroup(leader: &LeaderInfo, idx: usize, threads: &Vec<Arc<MyThread>>, main_sem: &sync::Semaphore) {
	let start = Instant::now();
	'outer: for t in threads {
		debug!( "Waiting for :{} ({})", t.id, t.idx.load(Ordering::Relaxed) );
		while t.idx.load(Ordering::Relaxed) != idx {
			hint::spin_loop();
			if start.elapsed() > Duration::from_secs(5) {
				eprintln!("Programs has been blocked for more than 5 secs");
				leader.estop.store(true, Ordering::Relaxed);
				main_sem.add_permits(1);
				break 'outer;
			}
		}
	}
	debug!( "Waiting done" );
}

fn wakegroup(exhaust: bool, me: usize, threads: &Vec<Arc<MyThread>>) {
	if !exhaust { return; }

	for i in 0..threads.len() {
		if i != me {
			debug!( "Leader waking {}", i);
			threads[i].sem.add_permits(1);
		}
	}
}

fn lead(exhaust: bool, leader: &LeaderInfo, this: & MyThread, threads: &Vec<Arc<MyThread>>, main_sem: &sync::Semaphore, exp: &bench::BenchData) {
	let nidx = leader.idx.load(Ordering::Relaxed) + 1;
	this.idx.store(nidx, Ordering::Relaxed);
	leader.idx.store(nidx, Ordering::Relaxed);

	if nidx as u64 > exp.stop_count || leader.estop.load(Ordering::Relaxed) {
		debug!( "Leader {} done", this.id);
		main_sem.add_permits(1);
		return;
	}

	debug!( "====================\nLeader no {} : {}", nidx, this.id);

	waitgroup(leader, nidx, threads, main_sem);

	leader.next( threads.len() );

	wakegroup(exhaust, this.id, threads);

	debug!( "Leader no {} : {} done\n====================", nidx, this.id);
}

async fn wait(exhaust: bool, leader: &LeaderInfo, this: &MyThread, rechecks: &mut usize) {
	task::yield_now().await;

	if leader.idx.load(Ordering::Relaxed) == this.idx.load(Ordering::Relaxed) {
		debug!("Waiting {} recheck", this.id);
		*rechecks += 1;
		return;
	}

	debug!("Waiting {}", this.id);

	debug_assert!( (leader.idx.load(Ordering::Relaxed) - 1) == this.idx.load(Ordering::Relaxed) );
	this.idx.fetch_add(1, Ordering::SeqCst);
	if exhaust {
		debug!("Waiting {} sem", this.id);
		this.sem.acquire().await.forget();
	}
	else {
		debug!("Waiting {} yield", this.id);
		task::yield_now().await;
	}

	debug!("Waiting {} done", this.id);
}

async fn transfer_main( me: usize, leader: Arc<LeaderInfo>, threads: Arc<Vec<Arc<MyThread>>>, exhaust: bool, start: Arc<sync::Barrier>, main_sem: Arc<sync::Semaphore>, exp: Arc<bench::BenchData>) -> usize{
	assert!( me == threads[me].id );

	debug!("Ready {}: {:p}", me, &threads[me].sem as *const sync::Semaphore);

	start.wait().await;

	debug!( "Start {}", me );

	let mut rechecks: usize = 0;

	loop {
		if leader.id.load(Ordering::Relaxed) == me {
			lead( exhaust, &leader, &threads[me], &threads, &main_sem, &exp );
			task::yield_now().await;
		}
		else {
			wait( exhaust, &leader, &threads[me], &mut rechecks ).await;
		}
		if leader.idx.load(Ordering::Relaxed) as u64 > exp.stop_count || leader.estop.load(Ordering::Relaxed) { break; }
	}

	rechecks
}

fn main() {
	let options = App::new("Transfer Tokio")
		.args(&bench::args())
		.arg(Arg::with_name("exhaust")  .short("e").long("exhaust")  .takes_value(true).default_value("no").help("Whether or not threads that have seen the new epoch should park instead of yielding."))
		.get_matches();

	let exhaust  = parse_yes_no( options.value_of("exhaust"), false );
	let nthreads = options.value_of("nthreads").unwrap().parse::<usize>().unwrap();
	let nprocs   = options.value_of("nprocs").unwrap().parse::<usize>().unwrap();


	let exp = Arc::new(bench::BenchData::new(options, nthreads, Some(100)));
	if exp.clock_mode {
		eprintln!("Programs does not support fixed duration mode");
		std::process::exit(1);
	}

	println!("Running {} threads on {} processors, doing {} iterations, {} exhaustion", nthreads, nprocs, exp.stop_count, if exhaust { "with" } else { "without" });

	let s = (1000000 as u64).to_formatted_string(&Locale::en);
	assert_eq!(&s, "1,000,000");

	let main_sem = Arc::new(sync::Semaphore::new(0));
	let leader = Arc::new(LeaderInfo::new(nthreads));
	let barr = Arc::new(sync::Barrier::new(nthreads + 1));
	let thddata : Arc<Vec<Arc<MyThread>>> = Arc::new(
		(0..nthreads).map(|i| {
			Arc::new(MyThread{
				id: i,
				idx: AtomicUsize::new(0),
				sem: sync::Semaphore::new(0),
			})
		}).collect()
	);

	let mut rechecks: usize = 0;
	let mut duration : std::time::Duration = std::time::Duration::from_secs(0);

	let runtime = Builder::new_multi_thread()
		.worker_threads(nprocs)
		.enable_all()
		.build()
		.unwrap();

	runtime.block_on(async {
		let threads: Vec<_> = (0..nthreads).map(|i| {
			tokio::spawn(transfer_main(i, leader.clone(), thddata.clone(), exhaust, barr.clone(), main_sem.clone(), exp.clone()))
		}).collect();
		println!("Starting");

		let start = Instant::now();

		barr.wait().await;
		debug!("Unlocked all");


		main_sem.acquire().await.forget();

		duration = start.elapsed();

		println!("\nDone");


		for i in 0..nthreads {
			thddata[i].sem.add_permits(1);
		}

		for t in threads {
			rechecks += t.await.unwrap();
		}
	});

	println!("Duration (ms)           : {}", (duration.as_millis()).to_formatted_string(&Locale::en));
	println!("Number of processors    : {}", (nprocs).to_formatted_string(&Locale::en));
	println!("Number of threads       : {}", (nthreads).to_formatted_string(&Locale::en));
	println!("Total Operations(ops)   : {:>15}", (leader.idx.load(Ordering::Relaxed) - 1).to_formatted_string(&Locale::en));
	println!("Threads parking on wait : {}", if exhaust { "yes" } else { "no" });
	println!("Rechecking              : {}", rechecks );
	println!("ns per transfer         : {}", ((duration.as_nanos() as f64) / leader.idx.load(Ordering::Relaxed) as f64));

}